File: MinMaxArray.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class MinMaxArray {

    public static List<Integer> findMinAndMax(int[] nums){
        List<Integer> result = new ArrayList<Integer>();
         int n=nums.length;


        int maxValue=nums[0];
        int minValue=nums[0];
        for(int i=1;i<n-1;i+=2){  //taking pairs -> reduced the no of comparisions
            if(nums[i]>nums[i+1]){
                maxValue=Integer.max(maxValue,nums[i]);
                minValue=Integer.min(minValue,nums[i+1]);
            }
            else{
                maxValue=Integer.max(maxValue,nums[i+1]);
                minValue=Integer.min(minValue,nums[i]);
            }
        }
        if(n%2==0){   
            maxValue=Integer.max(maxValue,nums[n-1]);
            minValue=Integer.min(minValue,nums[n-1]);
        }
        result.add(minValue);
        result.add(maxValue);

        return result;
    }
    public static void main(String[] args) {
        int[] arr1= new int[]{1,2,3,4,5};  //even
        int[] arr2 = new int[]{12,65,32,54,21,89,23,90}; //odd
        List<Integer> list1 = findMinAndMax(arr1);
        List<Integer> list2 = findMinAndMax(arr2);
        System.out.println("The minimum element of arr1 is: "+list1.get(0)
                            +" "+" and the maximum element of arr1 is: "+list1.get(1));
        System.out.println("The minimum element of arr2 is: "+list2.get(0)
                           +" "+" and the maximum element of arr1 is: "+list2.get(1));

    }
}

File: MinMaxArray.py
class Solution:
    """
    Time complexity - O(n). n - size of nums
    Space complexity - O(1)
    no of comparisions - 3 * (n/2)
    """
    def findMinMax(self, nums):
        result = []
        mini = 0
        maxi = 0
        if len(nums) % 2 == 0:
            mini = min(nums[0], nums[1])
            maxi = max(nums[0], nums[1])
            i = 2
        else:
            mini = nums[0]
            maxi = nums[0]
            i = 1
        while i < len(nums)-1:
            if nums[i] < nums[i+1]:
                mini = min(nums[i], mini)
                maxi = max(nums[i+1], maxi)
            else:
                mini = min(nums[i+1], mini)
                maxi = max(nums[i], maxi)
            i += 2
        print(" min - ", mini)
        print(" max - ", maxi)
    
def main():
    s =Solution()
    s.findMinMax([0,90,-90,100])
main()